### PR TITLE
[SERVICE-336] Clicking restore in the middle of another restore breaks restoration

### PR DESCRIPTION
**Pull request by bryangaleOF**
This still needs full testing, but would appreciate eyes on this.

There's four real parts to this

1. At the start and end of the restoreWorkspace method, mark and unmark that a restore is in progress, and if the restoreWorkspace  method is called again during a restore, throw an exception

2. Add a timeout when waiting for a  WorkspaceAPI.RESTORE_HANDLER call to the client app via sendToClient to return

3. Add a timeout when waiting for a client app to notify the provider that it is ready to restore

These 3 together /should/ be enough to ensure that we don't start a restore when we are already mid-restore but will also never block restores permanently due to a misbehaving client app. However, for caution, we also have:

4. A global timeout on the marker that restore is running, so that we definitely never permanently block restores

As part of this PR, I've also split up the restoreWorkspace to make it more manageable

### CODE REVIEW

File: src/provider/workspaces/restore.ts [hunk 1]
```diff
@@ -13,50 +13,98 @@ import {SCHEMA_MAJOR_VERSION} from './create';
 import {regroupWorkspace} from './group';
 import {addToWindowObject, childWindowPlaceholderCheck, childWindowPlaceholderCheckRunningApp, createNormalPlaceholder, createTabbedPlaceholderAndRecord, inWindowObject, parseVersionString, positionWindow, SemVer, TabbedPlaceholders, wasCreatedProgrammatically, WindowObject} from './utils';
 
+const GLOBAL_RESTORE_TIMEOUT = 120000;
+const CLIENT_STARTUP_TIMEOUT = 60000;
+const CLIENT_RESTORE_TIMEOUT = 60000;
```
Comment (pjbroadbent) [inline src/provider/workspaces/restore.ts]: Can you explain what each of these is used for? We also have a `Constants.ts`, which may be a better place for these.

File: src/provider/workspaces/restore.ts [hunk 2]
```diff
@@ -13,50 +13,98 @@ import {SCHEMA_MAJOR_VERSION} from './create';
...
+
+const appsCurrentlyStarting = new Map();
 const appsToRestore = new Map();
 const appsCurrentlyRestoring = new Map();
 
+let restoreBlocker: {}|null = null;
```
Comment (pjbroadbent) [inline src/provider/workspaces/restore.ts]: Can you explain the purpose of this.

Could we use the `Workspace` object as the token?

File: src/provider/workspaces/restore.ts [hunk 3]
```diff
@@ -13,50 +13,98 @@ import {SCHEMA_MAJOR_VERSION} from './create';
...
+export const restoreWorkspace = async(payload: Workspace, identity: Identity): Promise<Workspace> => {
+    if (restoreBlocker !== null) {
+        throw new Error('Attempting to restore while restore in progress');
+    }
+
+    validatePayload(payload);
+
+    startRestoreBlockerTimeout();
+
+    const layout = payload;
```
Comment (pjbroadbent) [inline src/provider/workspaces/restore.ts]: Whilst we're changing this code, can we start using "workspace" in favour of "layout"?

File: src/provider/workspaces/restore.ts [hunk 4]
```diff
@@ -13,50 +13,98 @@ import {SCHEMA_MAJOR_VERSION} from './create';
...
+    const startupApps: Promise<WorkspaceApp>[] = [];
+
+    await createAllPlaceholders(layout);
+
+    const apps = await promiseMap(layout.apps, async(app: WorkspaceApp): Promise<WorkspaceApp> => await restoreApp(app, startupApps));
```
Comment (pjbroadbent) [inline src/provider/workspaces/restore.ts]: The extra `async` and `await` here are a bit unnecessary, this line is equivalent to:
`const apps = await promiseMap(layout.apps, (app: WorkspaceApp): Promise<WorkspaceApp> => restoreApp(app, startupApps));`

For trivial lamba-type functions like this one, can probably omit the type declarations entirely?
`const apps = promiseMap(layout.apps, app => restoreApp(app, startupApps));`

File: src/provider/workspaces/restore.ts (hunk 1 above)
Comment (bryangaleOF) [reply src/provider/workspaces/restore.ts]: Constants.ts is under snapanddock, so doesn't feel like the right place, but have commented these constants

File: src/provider/workspaces/restore.ts (hunk 3 above)
Comment (bryangaleOF) [reply src/provider/workspaces/restore.ts]: Anticipating merge hell with SR-341 as is, so would prefer to leave this until after that's in

File: src/provider/workspaces/restore.ts (hunk 2 above)
Comment (bryangaleOF) [reply src/provider/workspaces/restore.ts]: Have improved naming and commenting around this. Would prefer to leave implementation as is and not use Workspace token, as global knowledge is needed to reason about this being usable for this purpose

**Comment by bryangaleOF** [comment]
Probably easier for PR-269/SERVICE-341 to go in before this (CC @MichaelMCoates)

**Comment by bryangaleOF** [comment]
Jenkins now green for this

File: src/provider/workspaces/restore.ts [hunk 5]
```diff
@@ -13,50 +13,104 @@ import {SCHEMA_MAJOR_VERSION} from './create';
...
+    return layout;
 };
 
-export const restoreApplication = async(layoutApp: WorkspaceApp, resolve: Function): Promise<void> => {
+const requestClientRestoreApp = async(layoutApp: WorkspaceApp, resolve: Function): Promise<void> => {
     const {uuid} = layoutApp;
     const name = uuid;
-    const defaultResponse: WorkspaceApp = {...layoutApp, childWindows: []};
     const appConnection = apiHandler.isClientConnection({uuid, name});
     if (appConnection) {
```
Comment (tomrobertsOF) [inline src/provider/workspaces/restore.ts]: Not part of your change, but here we assign to some consts and then only use them once in the following lines. Can this just be consolidated to something like
```
const {uuid} = layoutApp
if (apiHandler.isClientConnection({uuid: , name: uuid}) { ...
```

File: src/provider/workspaces/restore.ts [hunk 6]
```diff
@@ -85,9 +139,26 @@ export const restoreWorkspace = async(payload: Workspace, identity: Identity): P
...
+
+                appsCurrentlyStarting.clear();
+                appsToRestore.clear();
+                appsCurrentlyRestoring.clear();
+            }
+        }, GLOBAL_EXCLUSIVITY_TIMEOUT);
+    })();
+};
+
+const createAllPlaceholders = async(layout: Workspace): Promise<void> => {
```
Comment (tomrobertsOF) [inline src/provider/workspaces/restore.ts]: This function called `createAllPlaceholders` has an internally defined function also called `createAllPlaceholders`. May cause some confusion depending on scope. Maybe make the internal one `createPlaceholdersInternal` or something like that? Could possibly even just put it as an anonymous function inside the promiseMap.

File: src/provider/workspaces/restore.ts [hunk 7]
```diff
@@ -158,89 +228,102 @@ export const restoreWorkspace = async(payload: Workspace, identity: Identity): P
...
-                if (ofAppNotRunning) {
-                    await ofAppNotRunning.run().catch(console.log);
-                    const ofWindowNotRunning = await ofAppNotRunning.getWindow();
-                    await positionWindow(app.mainWindow);
-                }
-                // SHOULD WE RETURN DEFAULT RESPONSE HERE?!?
-                return defaultResponse;
+            if (ofAppNotRunning) {
+                await ofAppNotRunning.run().catch(console.log);
+                const ofWindowNotRunning = await ofAppNotRunning.getWindow();
```
Comment (tomrobertsOF) [inline src/provider/workspaces/restore.ts]: This is assigned but never used.

File: src/provider/workspaces/restore.ts [hunk 8]
```diff
@@ -158,89 +228,102 @@ export const restoreWorkspace = async(payload: Workspace, identity: Identity): P
...
-                    // If app created by manifest
-                    const {manifestUrl} = app;
-                    console.log('App has manifestUrl:', app);
-                    ofAppNotRunning = await fin.Application.createFromManifest(manifestUrl);
+                // Not connected to service
+                await positionWindow(app.mainWindow);
+                return defaultResponse;
+            }
+        } else {
+            let ofAppNotRunning: undefined|Application;
```
Comment (tomrobertsOF) [inline src/provider/workspaces/restore.ts]: More a matter of test, but can we make this `Application|undefined`

File: src/provider/workspaces/restore.ts [hunk 9]
```diff
@@ -158,89 +228,102 @@ export const restoreWorkspace = async(payload: Workspace, identity: Identity): P
...
+    } catch (e) {
+        console.error('Error restoring app', app, e);
+        return defaultResponse;
     }
+};
 
-    apiHandler.sendToAll('workspace-restored', layout);
+const clientRestoreAppWithTimeout = async(app: WorkspaceApp, mayBeLegacyApp: boolean): Promise<WorkspaceApp> => {
+    const {uuid} = app;
+    const name = uuid;
```
Comment (tomrobertsOF) [inline src/provider/workspaces/restore.ts]: uuid and name are only used as a pair. Can we consolidate as something like
`const mainWindowIdentity = {uuid: app.uuid, name: app.uuid}`
and change the below to just
`sendToClient<...>(mainWindowIdentity, ...)`

**Comment by MichaelMCoates** [comment]
All tests passed except for lint. Pushed up a lint commit. Will squash and merge now.

**Pull request merged**